### ISSUE TITLE
python3 __nonzero__ compatibility (Trac #2341)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2333]
Running muon level3 processing using python 3 (py3-v4) results in more events passing the cuts compared to using running it with python 2 (py2-v3.1.1).
The problem is the `CleanInputStreams` traysegment in `level3_Functions.py`, especially the following part (lines 79-90):

```text
def muon_stream(frame):
    if frame.Has("FilterMask"):
        for filter in ["MuonFilter_12", "MuonFilter_13"]:
            if frame["FilterMask"].has_key(filter) and bool(frame["FilterMask"][filter]):
                return True
    else:
        return False
    if frame.Has("QFilterMask"):
        for filter in ["EHEFilter_12", "EHEFilter_13"]:
            if frame["QFilterMask"].has_key(filter) and bool(frame["QFilterMask"][filter]):
                return True
    return False
```

`bool(frame["FilterMask"][filter])` will always evaluate to `True` in python 3, regardless of whether the filter mask is actually true or not. This results in events not passing the muon filter to be passed on for level3 processing.

Additionally, there is one print statement in `level3_Master.py` line 219 that is not yet python 3 compatible.

Migrated from https://code.icecube.wisc.edu/ticket/2341
```json
{
    "status": "closed", 
    "changetime": "2019-08-20T16:50:37", 
    "description": "Running muon level3 processing using python 3 (py3-v4) results in more events passing the cuts compared to using running it with python 2 (py2-v3.1.1).\nThe problem is the `CleanInputStreams` traysegment in `level3_Functions.py`, especially the following part (lines 79-90):\n{{{\ndef muon_stream(frame):\n    if frame.Has(\"FilterMask\"):\n        for filter in [\"MuonFilter_12\", \"MuonFilter_13\"]:\n            if frame[\"FilterMask\"].has_key(filter) and bool(frame[\"FilterMask\"][filter]):\n                return True\n    else:\n        return False\n    if frame.Has(\"QFilterMask\"):\n        for filter in [\"EHEFilter_12\", \"EHEFilter_13\"]:\n            if frame[\"QFilterMask\"].has_key(filter) and bool(frame[\"QFilterMask\"][filter]):\n                return True\n    return False\n}}}\n\n`bool(frame[\"FilterMask\"][filter])` will always evaluate to `True` in python 3, regardless of whether the filter mask is actually true or not. This results in events not passing the muon filter to be passed on for level3 processing.\n\nAdditionally, there is one print statement in `level3_Master.py` line 219 that is not yet python 3 compatible.", 
    "reporter": "eganster", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1566319837741391", 
    "component": "combo core", 
    "summary": "python3 __nonzero__ compatibility", 
    "priority": "critical", 
    "keywords": "", 
    "time": "2019-08-08T13:40:34", 
    "milestone": "Autumnal Equinox 2019", 
    "owner": "kjmeagher", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20190808T14:38:35: kjmeagher commented:

> I3FilterResult's pybindings defines a `__nonzero__` attribute:
> https://code.icecube.wisc.edu/projects/icecube/browser/IceCube/meta-projects/combo/trunk/dataclasses/private/pybindings/I3FilterResult.cxx#L69
> 
> If that is not working, then we need to investigate that, as it might be a symptom of something with wider implications about the python3 switch.

**Comment by icecube-trac** [comment]
Trac update at 20190808T15:44:32: eganster commented:

> Im not sure, but doing e.g.
> 
> ```text
> print(frame["FilterMask"]["SunFilter_13"])
> print(bool(frame["FilterMask"]["SunFilter_13"]))
> ```
> yields to different results in py2 and py3.
> python2 gives:
> 
> ```text
> [ I3FilterResult : 0 , 0 ]
> False
> ```
> whereas python3 gives:
> 
> ```text
> [ I3FilterResult : 0 , 0 ]
> True
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190808T15:49:50: david.schultz commented:

> Python 3 uses `__bool__` instead of `__nonzero__`.  So yes, this broke.

**Comment by icecube-trac** [comment]
Trac update at 20190808T15:55:15: 
* david.schultz commented:

> Files that need fixing: 
> 
> ```text
> topsimulator/python/validationsuite.py:    def __nonzero__(self):
> icetray/private/pybindings/I3Bool.cxx:    .def("__nonzero__", i3bool_bool)
> icetray/private/pybindings/I3Int.cxx:    .def("__nonzero__", i3int_bool)
> dataclasses/RELEASE_NOTES:* implement ``__nonzero__`` for I3Double
> dataclasses/private/pybindings/I3Double.cxx:    .def("__nonzero__", i3double_bool)
> dataclasses/private/pybindings/I3PODHolder/I3UInt64.cxx:      .def("__nonzero__", I3UInt64_bool)
> dataclasses/private/pybindings/I3MCTree.cxx:        .def("__nonzero__", &I3MCTree::empty)
> dataclasses/private/pybindings/I3FilterResult.cxx:    .def("__nonzero__",I3FilterResult_bool)
> dataclasses/resources/test/test_I3FilterResult.py:        self.assertTrue(fr1.__nonzero__(), "this should be true.")
> ```


* david.schultz changed priority from "normal" to "critical"
* david.schultz changed type from "enhancement" to "defect"
* david.schultz changed owner from "" to "kjmeagher"
* david.schultz changed milestone from "" to "Autumnal Equinox 2019"
* david.schultz changed component from "analysis" to "icecube offline"
* david.schultz changed status from "new" to "assigned"
* david.schultz changed summary from "level3-filter-muon is not python 3 compatible" to "python3 __nonzero__ compatibility"

**Comment by icecube-trac** [comment]
Trac update at 20190820T16:50:37: 
* kjmeagher commented:

> In changeset 175121:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="175121"
> Add `__bool__` operator to pybindings for python3 compatibility.
> 
> Effected classes are: `I3Double`, `I3UInt32`, `I3MCTree`, `I3FilterResult`, 
> `I3Bool`, `I3Int`, and `ValidationReport`. This is nescessary because python2
> uses `__nonzero__` for the binary operator, but python3 changed to `__bool__`.
> All of these classes now have both operators defined, for compatibility with both 
> languages.
> 
> Unit tests were added for all effeced classes.
> 
> In addition, it was noticed that the logic for `I3MCTree` was incorrect. It was changed 
> so that `I3MCTree` will be false on an empty tree and true on a non-empty tree. 
> It had mistakenly been set to the opposite. 
> 
> Also `test_tree_traversal` was enabled for `I3MCTree` as it had been mistakenly disabled 
> due to a formatting error.
> 
> fixes #2333
> ```


* kjmeagher changed status from "assigned" to "closed"
* kjmeagher changed resolution from "" to "fixed"
